doc: some updates

gmarull opened the pull request:
Refer to each commit for more details.

gmarull committed:
doc: add cmake SPHINXOPTS_EXTRA option

Add a new CMake option to append options to SPHINXOPTS. This allows us
to easily extend default SPHINXOPTS. This patch also restores the "-j
auto" option in CI (now that we use a custom runner).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
ci: doc-build: use zephyr-runner-linux-x64-4xlarge

This should allow us to observe faster build times when using
parallelization.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: enable Sphinx tracebacks

When something goes wrong in Sphinx or extensions, we'll get a detailed
traceback, so we can easily debug issues in CI.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: update sphinx/docleaf requirements

Update Sphinx and docleaf. Note that Sphinx 7 is not an option because
rtd_theme requires <= 6 (see
https://github.com/readthedocs/sphinx_rtd_theme/issues/1463).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: move requirements to doc/

Zephyr scripts do not require documentation dependencies, so let's
move them from scripts/ to doc/.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

kartben commented on .github/workflows/doc-build.yml:
```diff
@@ -35,7 +35,7 @@ env:
   doc-build-html:
     name: "Documentation Build (HTML)"
-    runs-on: ubuntu-22.04
+    runs-on: zephyr-runner-linux-x64-4xlarge
```
it would be good to stick to ubuntu-22.04 as it's free -- was the reason here only to try and speed up the build, or something else?

gmarull replied:
I saw PDF built was moved to this runner, and wanted to try with HTML. Build seems much faster (6-8min vs. 14-18min). If too expensive to run, we can continue using the free runner.

kartben replied:
Ya makes sense, we can give it a try. 
@nashif, FYI.

nashif replied:
yes, that looks good, every minute counts :)

gmarull committed:
doc: add -W --keep-going to the default SPHINXOPTS list

Run Sphinx in warnings as error mode, but keep going so that all issues
are reported to the user.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
@michaeljones I see `docleaf 0.8.0 depends on Sphinx!=5.0.0, <6 and >=4.0`, is there any reason to not support 6/7 versions of Sphinx?

michaeljones commented:
> @michaeljones I see `docleaf 0.8.0 depends on Sphinx!=5.0.0, <6 and >=4.0`, is there any reason to not support 6/7 versions of Sphinx?

Not that I'm aware of. I think I copied the version specifier from Breathe when I started the Docleaf project and haven't revisited it. I'm not sure what the ideal is but happy to widen it. Should I just remove the upper limit? I think there have been conversations about that in Breathe. The upper limit can be seen as overly cautious and a blocker for updates that might work fine.

gmarull commented:
> > @michaeljones I see `docleaf 0.8.0 depends on Sphinx!=5.0.0, <6 and >=4.0`, is there any reason to not support 6/7 versions of Sphinx?
> 
> Not that I'm aware of. I think I copied the version specifier from Breathe when I started the Docleaf project and haven't revisited it. I'm not sure what the ideal is but happy to widen it. Should I just remove the upper limit? I think there have been conversations about that in Breathe. The upper limit can be seen as overly cautious and a blocker for updates that might work fine.

Unless breathe relies on APIs not available on 6/7, it should relax the version spec.

michaeljones commented:
@gmarull - I've just released docleaf 0.8.1 which removes the upper limit on Sphinx.

carlescufi reviewed (changes requested):
All looks good to me except `doc: move requirements to doc/`, I am not sure I agree with this one, since we consolidate all requirements in `scripts/`.

gmarull commented:
> All looks good to me except `doc: move requirements to doc/`, I am not sure I agree with this one, since we consolidate all requirements in `scripts/`.

Documentation is not a script, everything lives under `doc/`. Also there's no need to install doc dependencies by default (most people just consumes docs)

carlescufi commented:
> > All looks good to me except `doc: move requirements to doc/`, I am not sure I agree with this one, since we consolidate all requirements in `scripts/`.
> 
> Documentation is not a script, everything lives under `doc/`. Also there's no need to install doc dependencies by default (most people just consume docs)

That is true, but note how we have `requirements-compliance.txt` in `scripts/`, which is comparable?

gmarull commented:
> > > All looks good to me except `doc: move requirements to doc/`, I am not sure I agree with this one, since we consolidate all requirements in `scripts/`.
> > 
> > 
> > Documentation is not a script, everything lives under `doc/`. Also there's no need to install doc dependencies by default (most people just consume docs)
> 
> That is true, but note how we have `requirements-compliance.txt` in `scripts/`, which is comparable?

`scripts/ci/check_compliance.py` :-)

nashif commented on scripts/requirements.txt:
```diff
@@ -1,6 +1,5 @@
 -r requirements-base.txt
 -r requirements-build-test.txt
--r requirements-doc.txt
```
now doc requirements will be missing if you `pip install -r scripts/requirements.txt`, no?

gmarull replied:
Yes, it is intentional. People who want to build docs will need to install doc-related dependencies using `pip install -r doc/requirements.txt`.

carlescufi replied:
Agreed, @gmarull how about we move it to `doc/` but then keep it part of the whole `requirements.txt`?

kartben replied:
Ya I'm with @gmarull here -- there are actually many additional non-Python reqs that people need to install anyways (Doxygen, Graphviz, ... see https://docs.zephyrproject.org/latest/contribute/documentation/generation.html#installing-the-documentation-processors) so there's no point in slowing down the initial onboarding experience IMO.
